Remove dead function cleanup from mandatory inlining

zoecarver opened the pull request:
<!-- What's in this pull request? -->
This patch adds `MandatoryFunctionCleanup`, another mandatory transformation pass.  I originally tried to do this in `MandatoryCombine` but ran into issues when trying to remove functions. `MandatoryFunctionCleanup` shouldn't have those issues because it is a `SILModuleTransform`. Moving more code out of mandatory inlining will make things like SourceKit faster because they don't run these passes. I plan to move dead closure cleanup into `MandatoryFunctionCleanup` next. 

refs #28394 #28396 
cc @gottesmm (thanks for being patient with all these patches!)

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

LucianoPAlmeida commented on lib/SILOptimizer/Mandatory/MandatoryFunctionCleanup.cpp:
```diff
@@ -0,0 +1,89 @@
+
+class MandatoryFunctionCleanup final : public SILModuleTransform {
+
+  /// Determin if we should remove unused transparent function.
```
typo on `Determin`? :))

slavapestov commented on lib/SILOptimizer/Mandatory/MandatoryFunctionCleanup.cpp:
```diff
@@ -0,0 +1,89 @@
+/// functions or [in the future: closures] that have been inlined in the
+/// mandatory inliner. This is a seprate pass from mandatory inlining so that it
+/// can be controlled individually. This makes it faster for things like
+/// SourceKit to run the "bare-minimum" beacuse they don't care about dead
```
Typo: "because"

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please benchmark

zoecarver commented:
I should have put a comment here earlier. 

I talked with @gottesmm and I'm not sure that a pass like this one is necessary at all. Another option would be to only remove the dead function elimination from mandatory inlining and let the dead function elimination optimization pass handle the removal of dead functions. The only downside is that IRGen might be slower.

zoecarver commented:
I've updated this code to just remove the removal of dead function from mandatory inlining. That doesn't need to be done at `-Onone`. 

cc @gottesmm

zoecarver commented:
I ran the tests and `mandatoryinlining-wrongdebugscope.swift` was the only failure (seems like there should be more, though). Anyway, can someone re-trigger the bots?

LucianoPAlmeida commented:
Minor thing, maybe the PR title should be updated? :)

zoecarver commented:
@LucianoPAlmeida will do. I wasn't sure if I should make a new PR or not-- some of the metadata around this one might be slightly off.

slavapestov commented:
@swift-ci Please smoke test

slavapestov reviewed (approved):
Removing code is always good :)

slavapestov commented:
@swift-ci Please smoke test

zoecarver commented:
Somehow removing this bit of code caused a crash. I'll investigate.

gottesmm commented:
This LGTM! For those listening in, this is really something that IRGen and or other passes should handle. Also, it is one step towards being able to have mandatory inlining be a function pass making it easier to debug.

zoecarver commented:
I forgot to write down what test(s) crashed the compiler in time :/ I fixed all the failing tests on my machine but that may differ from what fails on the bots. If someone could trigger the bots again, that would be great (no rush, I know its Christmas for many).

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

zoecarver commented:
Any ideas about what the issue is here? It seems to crash when trying to build the standard library (but, only on OS X buildbot). I can't seem to reproduce the issue.

LucianoPAlmeida commented:
Not sure, but as far as I could see I think the crash occurs under specific arguments, as you could note on the logs right above the crash, there are the `Program arguments` for this specific build that crashes there, maybe you've already done it but, have you try to reproduce using them? 
Maybe you could reproduce that under those arguments :))

zoecarver commented:
@LucianoPAlmeida Thanks for the suggestion. I'll try with the arguments.

gottesmm commented:
@zoecarver You should try using the exact preset on the bot. That is just good for you to be able to do.

That being said. The problem here is we got rid of the analysis invalidation. Any function that we inlined into needs to have its function body invalidated.

gottesmm commented:
@zoecarver Like this: https://github.com/apple/swift/pull/28484/files#diff-d78427f7bb2ea4c8304bcb6ef758d7bfL1013.

zoecarver commented:
@gottesmm thanks for pointing out the issue. I'll build with the bot config next time. 

Sorry for the slow turn around on this issue. I'll have some time to work on it tomorrow and hopefully finish it up.

zoecarver commented:
Adding `invalidateAnalysis` got the stdlib build working, thanks again @gottesmm. 

BTW also got the preset config working on my machine so I should be able to debug build failures better now.

gottesmm commented:
Can you clean up the history plz.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -996,40 +995,10 @@ class MandatoryInlining : public SILModuleTransform {
-      // Okay, just erase the function from the module.
-      FuncBuilder.eraseFunction(&F);
+      if (!getOptions().DebugSerialization) {
+        invalidateAnalysis(&F, SILAnalysis::InvalidationKind::Everything);
```
This is not sufficient without the pass manager change I mentioned. The inlining here is done recursively so you are going to need to also invalidate functions that are inlined into.

Otherwise, we will hit this non-deterministically. I think you can probably create a sil test case that does this. Let me describe it. If I am right, can you add a test for this?

Consider:

```
sil [transparent] @f {
  call g()
}
sil h() {
  doSomethingThatEffectsDominance()
}
sil [transparent] g() {
  h()
}
```

I think the optimizer will visit this here. The key thing you need to do is run a pass that constructs Dom Trees. I would be fine with an addition of a new optimizer pass that just does that.

zoecarver replied:
Hmm, good catch. Even though this pass runs on each function recursively, won't all the functions touched be in the module, so they will eventually, unconditionally (except if without `DebugSerialization`), get invalidated? 

I'm trying to work out a test case but, I'm not sure I understand what you mean by `doSomethingThatEffectsDominance`.
<details>
<summary>Here's what I've got so far (not much more than what you posted)</summary>

```
sil @f : $@convention(thin) () -> () {
bb0:
  // function_ref g()
  %0 = function_ref @g : $@convention(thin) () -> ()
  %1 = apply %0() : $@convention(thin) () -> ()
  %2 = tuple ()
  return %2 : $()
}

sil @h : $@convention(thin) () -> Builtin.Int1 {
bb0:
  %0 = integer_literal $Builtin.Int1, 0
  br bb2
bb1:
  return %0 : $Builtin.Int1
bb2:
  br bb1
}

sil [transparent] @g : $@convention(thin) () -> () {
bb0:
  // function_ref h()
  %0 = function_ref @h : $@convention(thin) () -> Builtin.Int1
  %1 = apply %0() : $@convention(thin) () -> Builtin.Int1
  %2 = tuple ()
  return %2 : $()
}
```

</details>

zoecarver replied:
I added a validation test to this PR, let me know what you think.

gottesmm commented:
I talked with @zoecarver off line and he confirmed that when run by hand, the given validation test fails as we expected it would without the invalidation fix.

gottesmm commented:
This is still incorrect. Please do not invalidate all functions like this. Instead, I would suggest passing in a SmallVector or something like that and then put any functions you inline into in there. Pop each such function off the vector at the top level and invalidate as you do the popping.

zoecarver commented:
@gottesmm I updated it based on your suggestion. Thanks for the comments.

gottesmm reviewed (approved):
LGTM

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -784,7 +784,8 @@ runOnFunctionRecursively(SILOptFunctionBuilder &FuncBuilder,
                          ImmutableFunctionSet CurrentInliningSet,
-                         ClassHierarchyAnalysis *CHA) {
+                         ClassHierarchyAnalysis *CHA,
+                         SmallPtrSetImpl<SILFunction *> &changedFunctions) {
```
I wouldn’t use a PtrSet for this. My suggestion instead would be use a small vector and in the caller call sortUnique. Iterating over Hash table backed sets is a code smell.

zoecarver replied:
I think the performance is more relevant here but, I don't feel strongly and the performance difference is probably minimal. I'll update it.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -997,39 +1005,11 @@ class MandatoryInlining : public SILModuleTransform {
-      // Okay, just erase the function from the module.
-      FuncBuilder.eraseFunction(&F);
+    
+    if (getOptions().DebugSerialization) return;
```
Can you not put this on the same line? Did you run this through git-clang-format?

zoecarver replied:
Sorry, I forgot. Formatted.

gottesmm commented:
LGTM, lets test this sucker

gottesmm commented:
Or actually, @zoecarver can you squash and then I will start tests?

zoecarver committed:
Removes dead function cleanup from MandatoryInlining.

This will help reduce compile time in SourceKit (among others). Dead
function elimination will handle the removal of dead functions (not in
-Onone).

zoecarver commented:
Squashed. @gottesmm thanks for all the help with this patch :)

gottesmm commented:
No worries! What is important is that we got to a good result.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci benchmark

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci benchmark

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
@swift-ci test source compatibility

zoecarver commented:
@shahmishal Any chance there's a swift docker image I could use to debug this?

futurejones commented:
Can one of the admins verify this patch?

zoecarver commented:
> Can one of the admins verify this patch?

That's not necessary. Currently, the hold is on my end. I need to debug the Linux failures (I think I might still be missing a case where I need to invalidate analysis).

zoecarver commented:
I wasn't able to make a reproducer but I think the issue may have occurred because devirtualized functions that weren't inlined weren't added to `changedFunctions`. Could someone re-trigger the Linux bots, or should I try to find an actual reproducer?

zoecarver commented:
@swift-ci please smoke test

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
This appears to be an unrelated lldb issue. 

@swift-ci please smoke test Linux.

zoecarver commented:
@swift-ci please smoke test OS X.

zoecarver commented:
@swift-ci please test Windows.

zoecarver commented:
@swift-ci please smoke test OS X.

zoecarver commented:
@swift-ci please test Windows.

zoecarver commented:
@swift-ci please smoke test Linux.

zoecarver commented on test/DebugInfo/mandatoryinlining-wrongdebugscope.swift:
```diff
@@ -0,0 +1,22 @@
+// RUN: %target-swift-frontend -emit-sil %s -Onone -Xllvm \
```
I was just about to ask if I can remove this test because it's not really testing anything... but it looks like @gottesmm beat me to it :))

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test OS X.

zoecarver commented:
@swift-ci please test Windows.

zoecarver commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -1030,38 +1041,11 @@ class MandatoryInlining : public SILModuleTransform {
-
-      // Okay, just erase the function from the module.
-      FuncBuilder.eraseFunction(&F);
+    sortUnique(changedFunctions);
```
Oops. Maybe `changedFunctions` should really be a set.

zoecarver replied:
Fixed.

zoecarver commented:
I'm going to do a source-compatibility test then merge this.

zoecarver commented:
@swift-ci please test.

zoecarver commented:
@swift-ci please test source compatibility.

zoecarver commented:
@swift-ci please test source compatibility.

zoecarver commented:
@swift-ci please test.

zoecarver commented:
In case anyone missed my earlier comment and has an objection: I'm planning to merge this tomorrow morning. 

(Assuming the approvals still apply but wanted to double-check given that it sat dormant for quite a while.)
